google-cloud-clients/google-cloud-firestore/src/main/java/com/google/cloud/firestore/FirestoreImpl.java

```diff
@@ -352,7 +352,25 @@ public void onSuccess(List<WriteResult> writeResults) {
                   @Override
                   public void run() {
                     try {
-                      callbackResult.set(transactionCallback.updateCallback(transaction));
+                      T callbackValue = transactionCallback.updateCallback(transaction);
+                      if (callbackValue instanceof ApiFuture) {
```

Comment:
This looks wrong.  if `T` is an `ApiFuture`, then it's not an `ApiFuture<T>`.  I would ask that we start over with a design for `AsyncFunction`